src/stp/computations/PassComputations.cpp

```diff
@@ -13,43 +13,93 @@
+    possibleReceiverLocations.reserve(us.size());
+    for (auto& robot : us) {
+        possibleReceiverLocations.emplace_back(robot->getPos());
     }
-    constexpr double MINIMUM_PASS_DISTANCE = 1.0;   // This can be dribbled instead of passed
-    constexpr double MINIMUM_LINE_OF_SIGHT = 10.0;  // The minimum LoS to be a valid pass, otherwise, the pass will go into an enemy robot
-    // TODO: create a better passingGrid (more points and/or better spread)
-    auto passGrid = Grid(-gridLength / 2, -gridWidth / 2, gridWidth, gridLength, numPoints, numPoints);  // 81 points spread over the whole field
-    for (auto& pointVector : passGrid.getPoints()) {
+
+    auto possiblePassLocationsVector = getPassGrid(field).getPoints();
+    for (auto& pointVector : possiblePassLocationsVector) {
         for (auto& point : pointVector) {
-            if (point.dist(ballLocation) < MINIMUM_PASS_DISTANCE) continue;
-            if (!FieldComputations::pointIsValidPosition(field, point, 2 * control_constants::ROBOT_RADIUS)) continue;
-            if (PositionScoring::scorePosition(point, gen::LineOfSight, field, world).score < MINIMUM_LINE_OF_SIGHT) continue;
-            for (auto robotLocation : robotLocations) {
-                // Robot travel time needs to be smaller than ball travel time in order to be able to receive the pass
-                if (calculateRobotTravelTime(robotLocation, point) < calculateBallTravelTime(ballLocation, passerLocation, point)) {
-                    gen::ScoredPosition scoredPosition = PositionScoring::scorePosition(point, profile, field, world);
-                    if (scoredPosition.score > bestPassLocation.score) {
-                        bestPassLocation = scoredPosition;
-                    }
-                    break;
+            if (pointIsValidPassLocation(point, ballLocation, possibleReceiverLocations, passerLocation, field, world)) {
+                gen::ScoredPosition scoredPosition = PositionScoring::scorePosition(point, profile, field, world);
+                if (scoredPosition.score > passInfo.passScore) {
+                    passInfo.passScore = scoredPosition.score;
+                    passInfo.passLocation = scoredPosition.position;
+                    passInfo.receiverId = world->getWorld()->getRobotClosestToPoint(passInfo.passLocation, us).value()->getId();
                 }
             }
         }
     }
-    if (bestPassLocation.score == 0) {
-        RTT_DEBUG("No good pass found! Passing to robot furthest in field.")
-        bestPassLocation.position = *std::max_element(robotLocations.begin(), robotLocations.end(), [](auto& p1, auto& p2) { return p1.x > p2.x; });
+
+    if (passInfo.passScore == 0) {
+        // If no good pass is found, pass to the robot furthest in the field
+        passInfo.passLocation = *std::max_element(possibleReceiverLocations.begin(), possibleReceiverLocations.end(), [](auto& p1, auto& p2) { return p1.x > p2.x; });
```

Comment:
As far as I can tell, there is no guarantee that `possibleReceiverLocations` has at least 1 location. If that is indeed the case, std::max_element might return an iterator that points to the end of the vector. Because you immediately dereference it, this code might crash. You could solve this in 2 ways:
- Either write a small if statement that checks if the return value of the max_element is a valid iterator, or
- Write in the comments of the function that there is a minimum required robots, so other people know when not to call this function

Reply:
Good catch. Added a check to make sure possibleReceviers won't be empty by the time it gets there.